platform/migration/platform-migration/src/main/java/org/codice/ddf/configuration/admin/ImportMigrationConfigurationAdminContext.java

```diff
@@ -0,0 +1,315 @@
+
+    /**
+     * Keeps track of all managed services in memory that were not found in the export file so
+     * we know what to delete at the end.
+     */
+    private final Map<String, Configuration> memoryServices;
+
+    /**
+     * Keeps track of all managed service factories in memory that were not found in the export file
+     * so we know what to delete at the end.
+     */
+    private final Map<String, List<Configuration>> memoryFactoryServices;
+
+    private final Map<String, ImportMigrationConfigurationAdminEntry> exportedServices;
+
+    private final Map<String, List<ImportMigrationConfigurationAdminEntry>> exportedFactoryServices;
+
+    private boolean isValid = true; // until proven otherwise
+
+    public ImportMigrationConfigurationAdminContext(ImportMigrationContext context,
+            ConfigurationAdminMigratable admin, ConfigurationAdmin configurationAdmin,
+            Configuration[] memoryConfigs) {
+        super(context);
+        Validate.notNull(admin, "invalid null configuration admin migratable");
+        Validate.notNull(configurationAdmin, "invalid null configuration admin");
+        Validate.notNull(memoryConfigs, "invalid null configurations");
+        this.admin = admin;
+        this.configurationAdmin = configurationAdmin;
+        // categorize memory configurations
+        this.memoryServices = Stream.of(memoryConfigs)
+                .filter(ImportMigrationConfigurationAdminContext::isManagedService)
+                .collect(Collectors.toMap(Configuration::getPid, Function.identity()));
+        this.memoryFactoryServices = Stream.of(memoryConfigs)
+                .filter(ImportMigrationConfigurationAdminContext::isManagedServiceFactory)
+                .collect(Collectors.groupingBy(Configuration::getFactoryPid));
+        // categorize exported configurations
+        final ImportMigrationConfigurationAdminEntry[] entries = context.entries()
+                .map(this::proxy)
+                .filter(Objects::nonNull)
+                .toArray(ImportMigrationConfigurationAdminEntry[]::new);
+
+        this.exportedServices = Stream.of(entries)
+                .filter(ImportMigrationConfigurationAdminEntry::isManagedService)
+                .collect(Collectors.toMap(ImportMigrationConfigurationAdminEntry::getPid,
+                        Function.identity()));
+        this.exportedFactoryServices = Stream.of(entries)
+                .filter(ImportMigrationConfigurationAdminEntry::isManagedServiceFactory)
+                .collect(Collectors.groupingBy(ImportMigrationConfigurationAdminEntry::getFactoryPid));
+        context.getReport()
+                .doAfterCompletion(this::deleteUnexportedConfigurationsAfterCompletion);
+    }
+
+    private static boolean isManagedServiceFactory(Configuration cfg) {
+        return cfg.getFactoryPid() != null;
+    }
+
+    private static boolean isManagedService(Configuration cfg) {
+        return cfg.getFactoryPid() == null;
```

Comment:
❗️ We should be checking the config's Dictionary for the factory-pid property. Either instead of the `getFactoryPid` method call or in addition to it.

Reply:
Based on our discussion ...
Your concern was about a known issue where a managed service factory cfg object was not properly created and we end up with the cfg object not reporting a factory pid but one gets entered in the properties. This creates a dummy cfg object with no managed service factory instance associated with it. As such, we will add protection on the export side to skip these cfg properties altogether.